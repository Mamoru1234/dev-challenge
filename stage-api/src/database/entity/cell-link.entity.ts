import {
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  RelationId,
} from 'typeorm';
import { CellEntity } from './cell.entity';

@Entity('cell-link')
export class CellLinkEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column('uuid')
  @RelationId((it: CellLinkEntity) => it.fromCell)
  fromCellId: string;

  @ManyToOne(() => CellEntity, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  fromCell: CellEntity;

  @Column('uuid')
  @RelationId((it: CellLinkEntity) => it.toCell)
  toCellId: string;

  @ManyToOne(() => CellEntity, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  toCell: CellEntity;
}
