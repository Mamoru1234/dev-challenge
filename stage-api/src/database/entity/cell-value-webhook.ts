import {
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  RelationId,
} from 'typeorm';
import { CellEntity } from './cell.entity';

@Entity('cell-value-webhook')
export class CellValueWebhookEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column('uuid')
  @RelationId((it: CellValueWebhookEntity) => it.cell)
  cellId: string;

  @ManyToOne(() => CellEntity, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  cell: CellEntity;

  @Column('text')
  url: string;
}
