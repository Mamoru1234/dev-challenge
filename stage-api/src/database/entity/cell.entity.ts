import { EquationNode } from 'equation-parser';
import {
  Column,
  Entity,
  Index,
  ManyToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { ExternalValueEntity } from './external-value.entity';

@Entity('cell')
@Index(['cellId', 'sheetId'])
export class CellEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column('text')
  cellId: string;

  @Column('text')
  @Index()
  sheetId: string;

  @Column('text')
  value: string;

  @Column('text')
  result: string;

  @Column({
    type: 'jsonb',
    nullable: true,
  })
  equation?: EquationNode;

  @ManyToMany(() => ExternalValueEntity, (value) => value.cells)
  externals: ExternalValueEntity[];
}
