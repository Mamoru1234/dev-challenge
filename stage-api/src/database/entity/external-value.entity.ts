import {
  Column,
  Entity,
  Index,
  JoinTable,
  ManyToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { CellEntity } from './cell.entity';

@Entity('external-value')
export class ExternalValueEntity {
  @PrimaryGeneratedColumn('increment')
  id: number;

  @Column('text')
  @Index({
    unique: true,
  })
  url: string;

  @Column('text')
  result: string;

  @ManyToMany(() => CellEntity, (value) => value.externals)
  @JoinTable()
  cells: CellEntity[];
}
