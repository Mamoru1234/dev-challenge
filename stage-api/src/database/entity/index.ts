import { CellLinkEntity } from './cell-link.entity';
import { CellValueWebhookEntity } from './cell-value-webhook';
import { CellEntity } from './cell.entity';
import { ExternalValueEntity } from './external-value.entity';

export const ENTITIES = [
  CellEntity,
  CellLinkEntity,
  CellValueWebhookEntity,
  ExternalValueEntity,
];
