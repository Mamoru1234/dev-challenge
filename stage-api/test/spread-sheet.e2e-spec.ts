import { Test, TestingModule } from '@nestjs/testing';
import { HttpStatus, INestApplication } from '@nestjs/common';
import request from 'supertest';
import { AppModule } from './../src/app.module';
import { getDataSourceToken } from '@nestjs/typeorm';
import { DataSource, Repository } from 'typeorm';
import { CellLinkEntity } from '../src/database/entity/cell-link.entity';
import { enableValidation } from '../src/app.utils';

describe.skip('Core spreadsheet API', () => {
  let app: INestApplication;
  let cellLinkRepository: Repository<CellLinkEntity>;
  const client = {
    getSheet: (sheetId: string) =>
      request(app.getHttpServer()).get(`/${sheetId}`),
    getCell: (sheetId: string, cellId: string) =>
      request(app.getHttpServer()).get(`/${sheetId}/${cellId}`),
    createCell: (sheetId: string, cellId: string, value: string) =>
      request(app.getHttpServer())
        .post(`/${sheetId}/${cellId}`)
        .send({ value }),
  };

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    enableValidation(app);
    await app.init();
    const dataSource: DataSource = app.get(getDataSourceToken('default'));
    cellLinkRepository = dataSource.getRepository(CellLinkEntity);
    await dataSource.query('TRUNCATE TABLE cell CASCADE;');
  });

  afterEach(async () => {
    app?.close();
  });

  it('should return not found if no cells', async () => {
    const { body } = await client.getSheet('test').expect(HttpStatus.NOT_FOUND);
    expect(body.message).toBe('sheet is missing');
  });

  async function createCells(sheetId: string, cells: [string, string][]) {
    for (const cell of cells) {
      await client
        .createCell(sheetId, cell[0], cell[1])
        .expect(HttpStatus.CREATED);
    }
  }

  it('should return value of single cell', async () => {
    await client.createCell('test', 'test_id', 'test_value');
    const { body } = await client.getSheet('test').expect(HttpStatus.OK);
    expect(body).toEqual({
      test_id: { value: 'test_value', result: 'test_value' },
    });
  });

  it('should return all values for this sheet only', async () => {
    await createCells('test', [
      ['test_id', 'test_value'],
      ['test_id_2', 'test_value_2'],
    ]);
    await client.createCell('test_2', 'test_id_3', 'test_value');
    const { body } = await client.getSheet('test').expect(HttpStatus.OK);
    expect(body).toEqual({
      test_id: { value: 'test_value', result: 'test_value' },
      test_id_2: {
        value: 'test_value_2',
        result: 'test_value_2',
      },
    });
  });

  it('simple case', async () => {
    await createCells('test', [
      ['var1', '2'],
      ['var2', '3'],
      ['var3', '=var1+var2'],
    ]);
    const { body } = await client.getSheet('test').expect(HttpStatus.OK);
    expect(body).toMatchSnapshot();
  });

  it('complicated case', async () => {
    await createCells('test', [
      ['a', '2'],
      ['b', '3'],
      ['c', '=a+b'],
      ['d', '=b*a'],
      ['e', '=c + d'],
    ]);
    const { body } = await client.getSheet('test').expect(HttpStatus.OK);
    expect(body).toMatchSnapshot();
    const links = await cellLinkRepository.count();
    expect(links).toBe(6);
  });

  it('simple update case', async () => {
    await createCells('test', [
      ['a', '2'],
      ['b', 'test'],
      ['a', 'test'],
      ['b', '3'],
    ]);
    const { body } = await client.getSheet('test').expect(HttpStatus.OK);
    expect(body).toMatchSnapshot();
  });

  it('simple dendency update', async () => {
    await createCells('test', [
      ['a', '2'],
      ['b', '3'],
      ['c', '=a+b'],
    ]);
    const { body: beforeUpdate } = await client
      .getSheet('test')
      .expect(HttpStatus.OK);
    expect(beforeUpdate).toMatchSnapshot();
    await client.createCell('test', 'a', '7').expect(HttpStatus.CREATED);
    const { body: afterUpdate } = await client
      .getSheet('test')
      .expect(HttpStatus.OK);
    expect(afterUpdate).toMatchSnapshot();
  });

  it('complex update case', async () => {
    await createCells('test', [
      ['a', '2'],
      ['b', '3'],
      ['c', '=a+b'],
      ['d', '=b*a'],
      ['e', '=c + d'],
      ['a', '4'],
    ]);
    const { body } = await client.getSheet('test').expect(HttpStatus.OK);
    expect(body).toMatchSnapshot();
    const links = await cellLinkRepository.count();
    expect(links).toBe(6);
  });

  it('simple circular reference', async () => {
    await createCells('test', [
      ['a', '2'],
      ['b', '3'],
      ['a', '=b'],
    ]);
    const { body } = await client
      .createCell('test', 'b', '=a')
      .expect(HttpStatus.UNPROCESSABLE_ENTITY);
    expect(body).toMatchSnapshot();
    const { body: sheetData } = await client
      .getSheet('test')
      .expect(HttpStatus.OK);
    expect(sheetData).toMatchSnapshot();
  });

  it('complex circular reference', async () => {
    await createCells('test', [
      ['a', '2'],
      ['b', '=a + 1'],
      ['c', '=b + 2'],
      ['e', '=b + 3'],
      ['d', '=c + 4'],
      ['f', '=d + 5 + e'],
    ]);
    const { body } = await client
      .createCell('test', 'c', '=b + f')
      .expect(HttpStatus.UNPROCESSABLE_ENTITY);
    expect(body).toMatchSnapshot();
    const { body: sheetData } = await client
      .getSheet('test')
      .expect(HttpStatus.OK);
    expect(sheetData).toMatchSnapshot();
  });

  it('complex circular reference with different cases', async () => {
    await createCells('test', [
      ['apple', '2'],
      ['baNanA', '=aPPle + 1'],
      ['CANDLE', '=SUM(BANANA, 2)'],
      ['Eggplant', '=BANaNA + 3'],
      ['dummy', '=CANDLE + 4'],
      ['f', '=dummy + 5 + EggPlanT'],
    ]);
    const { body } = await client
      .createCell('test', 'CANDLe', '=baNanA + f')
      .expect(HttpStatus.UNPROCESSABLE_ENTITY);
    expect(body).toMatchSnapshot();
    const { body: sheetData } = await client
      .getSheet('test')
      .expect(HttpStatus.OK);
    expect(sheetData).toMatchSnapshot();
  });

  it('function names in cells', async () => {
    await createCells('test', [
      ['SUM', '2'],
      ['sumOfSum', '=SUM(sum, 3)'],
    ]);
    const { body: sheetData } = await client
      .getSheet('test')
      .expect(HttpStatus.OK);
    expect(sheetData).toMatchSnapshot();
  });

  it('should update deps', async () => {
    await createCells('test', [
      ['a', '2'],
      ['b', '=a + 1'],
      ['c', '=b + 2'],
      ['e', '=b + 3'],
      ['d', '=c + 4'],
      ['f', '=d + 5 + e'],
      ['c', '=e + 5'],
    ]);
    const { body: sheetData } = await client
      .getSheet('test')
      .expect(HttpStatus.OK);
    expect(sheetData).toMatchSnapshot();
  });

  it('self reference', async () => {
    await createCells('test', [
      ['a', '2'],
      ['b', '3'],
    ]);
    const { body } = await client
      .createCell('test', 'b', '=b+a')
      .expect(HttpStatus.UNPROCESSABLE_ENTITY);
    expect(body).toMatchSnapshot();
    const { body: sheetData } = await client
      .getSheet('test')
      .expect(HttpStatus.OK);
    expect(sheetData).toMatchSnapshot();
  });

  it('wrong reference value', async () => {
    await createCells('test', [['a', 'test']]);
    const { body } = await client
      .createCell('test', 'b', '=a')
      .expect(HttpStatus.CREATED);
    expect(body).toMatchSnapshot();
  });

  it('get single cell', async () => {
    await createCells('test', [['a', 'test']]);
    const { body } = await client.getCell('test', 'a').expect(HttpStatus.OK);
    expect(body).toMatchSnapshot();
  });

  it('cell not found', async () => {
    await createCells('test', [['a', 'test']]);
    const { body } = await client
      .getCell('test', 'b')
      .expect(HttpStatus.NOT_FOUND);
    expect(body).toMatchSnapshot();
  });

  it('should handle update to string', async () => {
    await createCells('test', [
      ['a', '2'],
      ['b', '3'],
      ['c', '=a+b'],
    ]);
    const { body: beforeUpdate } = await client
      .getSheet('test')
      .expect(HttpStatus.OK);
    expect(beforeUpdate).toMatchSnapshot();
    await client
      .createCell('test', 'a', 'number of persons: ')
      .expect(HttpStatus.CREATED);
    const { body: afterUpdate } = await client
      .getSheet('test')
      .expect(HttpStatus.OK);
    expect(afterUpdate).toMatchSnapshot();
  });

  it('should drop links after update', async () => {
    await createCells('test', [
      ['a', '2'],
      ['b', '3'],
      ['c', '=a+b'],
    ]);
    const linksCountBefore = await cellLinkRepository.count();
    expect(linksCountBefore).toBe(2);
    await client.createCell('test', 'c', '3').expect(HttpStatus.CREATED);
    const linksCountAfter = await cellLinkRepository.count();
    expect(linksCountAfter).toBe(0);
  });
});
