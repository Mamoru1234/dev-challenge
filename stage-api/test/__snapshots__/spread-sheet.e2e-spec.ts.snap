// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpreadSheetController (e2e) cell not found 1`] = `
{
  "error": "Not Found",
  "message": "Cell not found",
  "statusCode": 404,
}
`;

exports[`SpreadSheetController (e2e) comples update case 1`] = `
{
  "a": {
    "result": "4",
    "value": "4",
  },
  "b": {
    "result": "3",
    "value": "3",
  },
  "c": {
    "result": "7",
    "value": "=a+b",
  },
  "d": {
    "result": "12",
    "value": "=b*a",
  },
  "e": {
    "result": "19",
    "value": "=c + d",
  },
}
`;

exports[`SpreadSheetController (e2e) complex circular reference 1`] = `
{
  "error": "Unprocessable Entity",
  "message": "Circular reference found",
  "statusCode": 422,
}
`;

exports[`SpreadSheetController (e2e) complex circular reference 2`] = `
{
  "a": {
    "result": "2",
    "value": "2",
  },
  "b": {
    "result": "3",
    "value": "=a + 1",
  },
  "c": {
    "result": "5",
    "value": "=b + 2",
  },
  "d": {
    "result": "9",
    "value": "=c + 4",
  },
  "e": {
    "result": "6",
    "value": "=b + 3",
  },
  "f": {
    "result": "20",
    "value": "=d + 5 + e",
  },
}
`;

exports[`SpreadSheetController (e2e) complex circular reference with different cases 1`] = `
{
  "error": "Unprocessable Entity",
  "message": "Circular reference found",
  "statusCode": 422,
}
`;

exports[`SpreadSheetController (e2e) complex circular reference with different cases 2`] = `
{
  "apple": {
    "result": "2",
    "value": "2",
  },
  "banana": {
    "result": "3",
    "value": "=aPPle + 1",
  },
  "candle": {
    "result": "5",
    "value": "=SUM(BANANA, 2)",
  },
  "dummy": {
    "result": "9",
    "value": "=CANDLE + 4",
  },
  "eggplant": {
    "result": "6",
    "value": "=BANaNA + 3",
  },
  "f": {
    "result": "20",
    "value": "=dummy + 5 + EggPlanT",
  },
}
`;

exports[`SpreadSheetController (e2e) complicated case 1`] = `
{
  "a": {
    "result": "2",
    "value": "2",
  },
  "b": {
    "result": "3",
    "value": "3",
  },
  "c": {
    "result": "5",
    "value": "=a+b",
  },
  "d": {
    "result": "6",
    "value": "=b*a",
  },
  "e": {
    "result": "11",
    "value": "=c + d",
  },
}
`;

exports[`SpreadSheetController (e2e) function names in cells 1`] = `
{
  "sum": {
    "result": "2",
    "value": "2",
  },
  "sumofsum": {
    "result": "5",
    "value": "=SUM(sum, 3)",
  },
}
`;

exports[`SpreadSheetController (e2e) get single cell 1`] = `
{
  "result": "test",
  "value": "test",
}
`;

exports[`SpreadSheetController (e2e) self reference 1`] = `
{
  "error": "Unprocessable Entity",
  "message": "Self reference is not allowed",
  "statusCode": 422,
}
`;

exports[`SpreadSheetController (e2e) self reference 2`] = `
{
  "a": {
    "result": "2",
    "value": "2",
  },
  "b": {
    "result": "3",
    "value": "3",
  },
}
`;

exports[`SpreadSheetController (e2e) should update deps 1`] = `
{
  "a": {
    "result": "2",
    "value": "2",
  },
  "b": {
    "result": "3",
    "value": "=a + 1",
  },
  "c": {
    "result": "11",
    "value": "=e + 5",
  },
  "d": {
    "result": "15",
    "value": "=c + 4",
  },
  "e": {
    "result": "6",
    "value": "=b + 3",
  },
  "f": {
    "result": "26",
    "value": "=d + 5 + e",
  },
}
`;

exports[`SpreadSheetController (e2e) simple case 1`] = `
{
  "var1": {
    "result": "2",
    "value": "2",
  },
  "var2": {
    "result": "3",
    "value": "3",
  },
  "var3": {
    "result": "5",
    "value": "=var1+var2",
  },
}
`;

exports[`SpreadSheetController (e2e) simple circular reference 1`] = `
{
  "error": "Unprocessable Entity",
  "message": "Circular reference found",
  "statusCode": 422,
}
`;

exports[`SpreadSheetController (e2e) simple circular reference 2`] = `
{
  "a": {
    "result": "3",
    "value": "=b",
  },
  "b": {
    "result": "3",
    "value": "3",
  },
}
`;

exports[`SpreadSheetController (e2e) simple update case 1`] = `
{
  "a": {
    "result": "test",
    "value": "test",
  },
  "b": {
    "result": "3",
    "value": "3",
  },
}
`;

exports[`SpreadSheetController (e2e) wrong reference value 1`] = `
{
  "error": "Unprocessable Entity",
  "message": "a is referenced in formula but cannot be parsed",
  "statusCode": 422,
}
`;
